plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,941 @@
+
+        # Calculate normal vectors (facing towards the grid origin) for both
+        # the source and detector planes
+        self.src_n = -self.source / np.linalg.norm(self.source)
+        self.det_n = -self.detector / np.linalg.norm(self.detector)
+
+        # Vector directly from source to detector
+        self.src_det = self.detector - self.source
+
+        # Magnification
+        self.mag = 1 + np.linalg.norm(self.detector) / np.linalg.norm(self.source)
+        self._log(f"Magnification: {self.mag}")
+
+        # Check that source-detector vector actually passes through the grid
+        if not self.grid.vector_intersects(self.source * u.m, self.detector * u.m):
+            raise ValueError(
+                "The vector between the source and the detector "
+                "does not intersect the grid provided!"
+            )
+
+        # Determine the angle above which particles will not hit the grid
+        # these particles can be ignored until the end of the simulation,
+        # then immediately advanced to the detector grid with their original
+        # velocities
+        self.max_theta_hit_grid = self._max_theta_hit_grid()
+
+        # ************************************************************************
+        # Define the detector plane
+        # ************************************************************************
+
+        # Load or calculate the detector hdir
+        if detector_hdir is not None:
+            self.det_hdir = detector_hdir / np.linalg.norm(detector_hdir)
+        else:
+            self.det_hdir = self._default_detector_hdir()
+
+        # Calculate the detector vdir
+        ny = np.cross(self.det_hdir, self.det_n)
+        self.det_vdir = -ny / np.linalg.norm(ny)
+
+        # ************************************************************************
+        # Validate the E and B fields
+        # ************************************************************************
+
+        req_quantities = ["E_x", "E_y", "E_z", "B_x", "B_y", "B_z"]
+        for rq in req_quantities:
+
+            # Error check that grid contains E and B variables required
+            if rq not in self.grid.quantities:
+                warnings.warn(
+                    f"{rq} not specified for provided grid."
+                    "This quantity will be assumed to be zero.",
+                    RuntimeWarning,
+                )
+                # If missing, warn user and then replace with an array of zeros
+                unit = self.grid._recognized_quantities[rq].unit
+                arg = {rq: np.zeros(self.grid.shape) * unit}
+                self.grid.add_quantities(**arg)
+
+            # Check that there are no infinite values
+            if not np.isfinite(self.grid[rq]).all():
+                raise ValueError(
+                    f"Input arrays must be finite: {rq} contains "
+                    "either NaN or infinite values."
+                )
```

Comment:
Interesting question: maybe? I don't think we should necessarily prevent all grids from having infinite values or any set of required quantities. But I could add methods to `grid.AbstractGrid` that checks for these conditions and returns a Boolean, then use those functions here to raise the exceptions?

Reply:
Yeah, that's what I meant to and failed to communicate, sorry! :D

Reply:
Ok, so I just did this for checking required quantities (`grid.require_quantities`) since that seemed like a pretty common problem. 

Upon further reflection, I'm not convinced that checking for infinite values or large value boundaries deserve their own functions in grids yet? Just weighing the value of moving the code there vs. the cost of adding even more user-facing methods to clog up the documentation?